logger

steenzout opened the pull request:
- added `monocle.logger` module 
- added `monocle.logger.Adapter` class

@sah please review. thank you!

lukmdo commented on monocle/logger.py:
```diff
@@ -0,0 +1,37 @@
+        ex = sys.exc_info()[1]
+
+        if hasattr(ex, '_monocle'):
+            self.logger.error('%s\n%s', msg, format_tb(ex))
```
maybe worth adding `**kw` like `exc_info=None, extra=None` could be specified and would be gone 
I am specifically after `extra`

lukmdo commented:
nice work ! 

given `Adapter` needs `logger` how about adding convince method like `getLogger(*a)` to be used by thir parties.

I think now the will have to do
```py
import logging
from monocle.logger import Adapter 

log = logging.getLogger(__name__)
log = monocle.logger.Adapter(log)
```

what I shoot at would look transform above to
```
from monocle.logger import getLogger  

log = getLogger(__name__)
```

steenzout commented:
@lukmdo 
- added a `monocle.logger.get()`
- fixed the `monocle.logger.Adapter.exception()`; it now formats the log message in the same way as the `Logger.exception()` and just appends the monocle stack trace to the message; this should cover all the cases where someone adds extra `args` or `kwargs`

let me know if there is anything else I can do to improve this.

cheers!

steenzout commented:
in case you want to manually test the output here is a code sample:

```
import monocle
import monocle.logger

from monocle import _o
from monocle.stack import eventloop


logger = monocle.logger.get()

@_o
def t():
    @_o
    def e():
        raise Exception('Test')

    try:
        _ = yield e()
    except BaseException as e:
        logger.exception('%s %s', 'error', e)
        raise e
    finally:
        eventloop.halt()

if __name__ == '__main__':
    monocle.launch(t)
    eventloop.run()
```

sah commented on monocle/logger.py:
```diff
@@ -0,0 +1,50 @@
+from monocle.core import format_tb
+
+
+def get(name=None):
```
Let's call this `getLogger`, like the equivalent function in the Python logging module.

sah replied:
It might also make sense to call this module `logging`. We'll need to be careful in monocle code though with this shadowing the standard logging module.

sah commented on monocle/logger.py:
```diff
@@ -0,0 +1,50 @@
+        this function will write to this adapter's logger
+        the exception message with its stack trace.
+        Otherwise, it will delegate the log message processing to the underlying logger.
+        """
```
This comment should explain why this is desirable. I think the key fact to mention is that exceptions thrown from o-routines contain extra metadata to allow us to reconstruct a traceback that follows the stack through the round trips to the event loop that happen at every `yield` in an o-routine.

steenzout commented:
- `logger` renamed to `logging`
- took measures to prevent `monocle.logging` to shadow `logging`
- renamed `monocle.logging.get()` to `monocle.logging.getLogger()`
- docstring updated

Note: python2.6 builds will fail (something the latest versions of twisted don't support anymore) due to this:

```
super(Adapter, self).__init__(logger, None)
```

let me know if I really need to fix this or not.

lukmdo commented on monocle/logging.py:
```diff
@@ -0,0 +1,51 @@
+
+        :param logger: the logger instance.
+        """
+        super(Adapter, self).__init__(logger, None)
```
maybe better `extra=None` or the `*args, **kwargs` passing

lukmdo replied:
is that `__init__` really  needed ?

steenzout replied:
@lukmdo confirmed there is not need for the `__init__`:

```
>>> class A(object):
...     def __init__(self, x, y):
...         self.x = x
...         self.y = y
... 
>>> class B(A):
...     def func(self):
...         print '%s %s' % (self.x, self.y)
... 
>>> b = B()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: __init__() takes exactly 3 arguments (1 given)
>>> b = B(1, 2)
>>> b.func()
1 2
```

removing it.

lukmdo replied:
Thank you @steenzout  for providing that extra bit

However my doubt still is open if we need it. Especially:
- main gate to that functionality will be `getLogger` that can pass `extra=None` where needed.
- if Adapter needs to be used directly then still I see no need to change the original signature (therefore could advise to keep it).

steenzout replied:
if someone needs to use the `Adapter` and provide some contextual information they can use it by:

```
logger = logging.getLogger()
adapter = monocle.logging.Adapter(logger, extra={'x': 1})
```

[the Python docs example](https://docs.python.org/2/howto/logging-cookbook.html#context-info) seem to suggest that if you pass `extra` you'll also override the `process` function of the `Adapter` to prefix or append contextual information to the log message.
since we're not overriding `process` I don't think it makes sense to allow it to pass `extra` on the `getLogger` call; the effect will only be to add keys to the `kwargs` dictionary which is pointless.

if someone really needs the contextual information they can subclass `monocle.logging.Adapter` and override `process`.

we can always revisit this in the future if the need arises otherwise I'd just leave it as it is now.

steenzout replied:
github is having an outage.

my commit with the removal of `__init__` is not showing up...

lukmdo commented:
need @steenzout to OK dropping 2.5, 2.6 (maybe can workaround if that is just one `super`) 
If so that would probably also need 
- version bump 
- can check for `sys.version_info` like in setup.py

steenzout commented:
@lukmdo @sah, said it was ok to drop support for Python < 2.7.

steenzout commented:
@lukmdo almost 2 months :)
